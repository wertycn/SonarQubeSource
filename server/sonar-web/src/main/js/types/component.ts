/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Task } from './tasks';
import { Component, LightComponent } from './types';

export enum Visibility {
  Public = 'public',
  Private = 'private',
}

export enum ComponentQualifier {
  Application = 'APP',
  Directory = 'DIR',
  File = 'FIL',
  Portfolio = 'VW',
  Project = 'TRK',
  SubPortfolio = 'SVW',
  TestFile = 'UTS',
}

export enum ProjectKeyValidationResult {
  Valid = 'valid',
  Empty = 'empty',
  TooLong = 'too_long',
  InvalidChar = 'invalid_char',
  OnlyDigits = 'only_digits',
}

export interface TreeComponent extends LightComponent {
  id?: string;
  name: string;
  path?: string;
  refId?: string;
  refKey?: string;
  tags?: string[];
  visibility: Visibility;
}

export interface TreeComponentWithPath extends TreeComponent {
  path: string;
}

export function isPortfolioLike(
  componentQualifier?: string | ComponentQualifier,
): componentQualifier is ComponentQualifier.Portfolio | ComponentQualifier.SubPortfolio {
  return (
    componentQualifier === ComponentQualifier.Portfolio ||
    componentQualifier === ComponentQualifier.SubPortfolio
  );
}

export function isApplication(
  componentQualifier?: string | ComponentQualifier,
): componentQualifier is ComponentQualifier.Application {
  return componentQualifier === ComponentQualifier.Application;
}

export function isProject(
  componentQualifier?: string | ComponentQualifier,
): componentQualifier is ComponentQualifier.Project {
  return componentQualifier === ComponentQualifier.Project;
}

export function isFile(
  componentQualifier?: string | ComponentQualifier,
): componentQualifier is ComponentQualifier.File {
  return [ComponentQualifier.File, ComponentQualifier.TestFile].includes(
    componentQualifier as ComponentQualifier,
  );
}

export function isView(
  componentQualifier?: string | ComponentQualifier,
): componentQualifier is
  | ComponentQualifier.Application
  | ComponentQualifier.Portfolio
  | ComponentQualifier.SubPortfolio {
  return isPortfolioLike(componentQualifier) || isApplication(componentQualifier);
}

export interface ComponentContextShape {
  component?: Component;
  currentTask?: Task;
  isInProgress?: boolean;
  isPending?: boolean;
  onComponentChange: (changes: Partial<Component>) => void;
  fetchComponent: () => Promise<void>;
}
