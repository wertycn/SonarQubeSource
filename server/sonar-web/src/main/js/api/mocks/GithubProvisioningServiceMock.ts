/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockTask } from '../../helpers/mocks/tasks';
import {
  GitHubConfigurationStatus,
  GitHubMapping,
  GitHubProvisioningStatus,
} from '../../types/provisioning';
import { Task, TaskStatuses, TaskTypes } from '../../types/tasks';
import {
  activateGithubProvisioning,
  addGithubRolesMapping,
  checkConfigurationValidity,
  deactivateGithubProvisioning,
  deleteGithubRolesMapping,
  fetchGithubProvisioningStatus,
  fetchGithubRolesMapping,
  updateGithubRolesMapping,
} from '../github-provisioning';

jest.mock('../github-provisioning');

const defaultConfigurationStatus: GitHubConfigurationStatus = {
  application: {
    jit: {
      status: GitHubProvisioningStatus.Success,
    },
    autoProvisioning: {
      status: GitHubProvisioningStatus.Success,
    },
  },
  installations: [
    {
      organization: 'testOrg',
      autoProvisioning: {
        status: GitHubProvisioningStatus.Success,
      },
      jit: {
        status: GitHubProvisioningStatus.Success,
      },
    },
  ],
};

const githubMappingMock = (
  id: string,
  permissions: (keyof GitHubMapping['permissions'])[],
  baseRole = false,
) => ({
  id,
  githubRole: id,
  baseRole,
  permissions: {
    user: permissions.includes('user'),
    codeViewer: permissions.includes('codeViewer'),
    issueAdmin: permissions.includes('issueAdmin'),
    securityHotspotAdmin: permissions.includes('securityHotspotAdmin'),
    admin: permissions.includes('admin'),
    scan: permissions.includes('scan'),
  },
});

const defaultMapping: GitHubMapping[] = [
  githubMappingMock('read', ['user', 'codeViewer'], true),
  githubMappingMock(
    'write',
    ['user', 'codeViewer', 'issueAdmin', 'securityHotspotAdmin', 'scan'],
    true,
  ),
  githubMappingMock('triage', ['user', 'codeViewer'], true),
  githubMappingMock(
    'maintain',
    ['user', 'codeViewer', 'issueAdmin', 'securityHotspotAdmin', 'scan'],
    true,
  ),
  githubMappingMock(
    'admin',
    ['user', 'codeViewer', 'issueAdmin', 'securityHotspotAdmin', 'admin', 'scan'],
    true,
  ),
];

export default class GithubProvisioningServiceMock {
  githubProvisioningStatus: boolean;
  githubConfigurationStatus: GitHubConfigurationStatus;
  githubMapping: GitHubMapping[];
  tasks: Task[];

  constructor() {
    this.githubProvisioningStatus = false;
    this.githubConfigurationStatus = cloneDeep(defaultConfigurationStatus);
    this.githubMapping = cloneDeep(defaultMapping);
    this.tasks = [];
    jest
      .mocked(activateGithubProvisioning)
      .mockImplementation(this.handleActivateGithubProvisioning);
    jest
      .mocked(deactivateGithubProvisioning)
      .mockImplementation(this.handleDeactivateGithubProvisioning);
    jest
      .mocked(fetchGithubProvisioningStatus)
      .mockImplementation(this.handleFetchGithubProvisioningStatus);
    jest
      .mocked(checkConfigurationValidity)
      .mockImplementation(this.handleCheckConfigurationValidity);
    jest.mocked(fetchGithubRolesMapping).mockImplementation(this.handleFetchGithubRolesMapping);
    jest.mocked(updateGithubRolesMapping).mockImplementation(this.handleUpdateGithubRolesMapping);
    jest.mocked(addGithubRolesMapping).mockImplementation(this.handleAddGithubRolesMapping);
    jest.mocked(deleteGithubRolesMapping).mockImplementation(this.handleDeleteGithubRolesMapping);
  }

  addProvisioningTask = (overrides: Partial<Omit<Task, 'type'>> = {}) => {
    this.tasks.push(
      mockTask({
        id: Math.random().toString(),
        type: TaskTypes.GithubProvisioning,
        ...overrides,
      }),
    );
  };

  setConfigurationValidity = (overrides: Partial<GitHubConfigurationStatus> = {}) => {
    this.githubConfigurationStatus = {
      ...this.githubConfigurationStatus,
      ...overrides,
    };
  };

  enableGithubProvisioning = () => {
    this.githubProvisioningStatus = true;
  };

  handleActivateGithubProvisioning = () => {
    this.githubProvisioningStatus = true;
    return Promise.resolve();
  };

  handleDeactivateGithubProvisioning = () => {
    this.githubProvisioningStatus = false;
    return Promise.resolve();
  };

  handleFetchGithubProvisioningStatus = () => {
    if (!this.githubProvisioningStatus) {
      return Promise.resolve({ enabled: false });
    }

    const nextSync = this.tasks.find((t: Task) =>
      [TaskStatuses.InProgress, TaskStatuses.Pending].includes(t.status),
    );
    const lastSync = this.tasks.find(
      (t: Task) => ![TaskStatuses.InProgress, TaskStatuses.Pending].includes(t.status),
    );

    return Promise.resolve({
      enabled: true,
      nextSync: nextSync ? { status: nextSync.status } : undefined,
      lastSync: lastSync
        ? {
            status: lastSync.status,
            finishedAt: lastSync.executedAt,
            startedAt: lastSync.startedAt,
            executionTimeMs: lastSync.executionTimeMs,
            summary: lastSync.status === TaskStatuses.Success ? 'Test summary' : undefined,
            errorMessage: lastSync.errorMessage,
            warningMessage: lastSync.warnings?.join() ?? undefined,
          }
        : undefined,
    });
  };

  handleCheckConfigurationValidity = () => {
    return Promise.resolve(this.githubConfigurationStatus);
  };

  handleFetchGithubRolesMapping: typeof fetchGithubRolesMapping = () => {
    return Promise.resolve(this.githubMapping);
  };

  handleUpdateGithubRolesMapping: typeof updateGithubRolesMapping = (id, data) => {
    this.githubMapping = this.githubMapping.map((mapping) =>
      mapping.id === id ? { ...mapping, ...data } : mapping,
    );

    return Promise.resolve(
      this.githubMapping.find((mapping) => mapping.id === id) as GitHubMapping,
    );
  };

  handleAddGithubRolesMapping: typeof addGithubRolesMapping = (data) => {
    const newRole = { ...data, id: data.githubRole };
    this.githubMapping = [...this.githubMapping, newRole];

    return Promise.resolve(newRole);
  };

  handleDeleteGithubRolesMapping: typeof deleteGithubRolesMapping = (id) => {
    this.githubMapping = this.githubMapping.filter((el) => el.id !== id);
    return Promise.resolve();
  };

  addGitHubCustomRole = (id: string, permissions: (keyof GitHubMapping['permissions'])[]) => {
    this.githubMapping = [...this.githubMapping, githubMappingMock(id, permissions)];
  };

  reset = () => {
    this.githubProvisioningStatus = false;
    this.githubConfigurationStatus = cloneDeep(defaultConfigurationStatus);
    this.githubMapping = cloneDeep(defaultMapping);
    this.tasks = [];
  };
}
