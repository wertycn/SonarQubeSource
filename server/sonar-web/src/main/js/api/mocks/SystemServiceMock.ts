/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { Provider, SysInfoCluster, SysInfoLogging, SysInfoStandalone } from '../../types/types';

import { LogsLevels } from '../../apps/system/utils';
import { mockClusterSysInfo, mockLogs, mockStandaloneSysInfo } from '../../helpers/testMocks';
import { getSystemInfo, setLogLevel } from '../system';

jest.mock('../system');

export default class SystemServiceMock {
  isCluster: boolean = false;
  logging: SysInfoLogging = mockLogs();
  systemInfo: SysInfoCluster | SysInfoStandalone = mockStandaloneSysInfo();

  constructor() {
    this.updateSystemInfo();
    jest.mocked(getSystemInfo).mockImplementation(this.handleGetSystemInfo);
    jest.mocked(setLogLevel).mockImplementation(this.handleSetLogLevel);
  }

  handleGetSystemInfo = () => {
    return this.reply(this.systemInfo);
  };

  setProvider(provider: Provider | null) {
    this.systemInfo = mockStandaloneSysInfo({
      ...this.systemInfo,
      System: {
        ...this.systemInfo.System,
        ...(provider ? { 'External Users and Groups Provisioning': provider } : {}),
      },
    });
  }

  handleSetLogLevel = (logsLevel: LogsLevels) => {
    this.logging = mockLogs(logsLevel);
    this.updateSystemInfo();

    return this.reply(undefined);
  };

  updateSystemInfo = () => {
    const logs = {
      'Web Logging': this.logging,
      'Compute Engine Logging': this.logging,
    };

    this.systemInfo = this.isCluster ? mockClusterSysInfo(logs) : mockStandaloneSysInfo(logs);
  };

  setIsCluster = (isCluster: boolean = false) => {
    this.isCluster = isCluster;
    this.updateSystemInfo();
  };

  reset = () => {
    this.logging = mockLogs();
    this.setIsCluster(false);
    this.updateSystemInfo();
  };

  reply<T>(response: T): Promise<T> {
    return Promise.resolve(cloneDeep(response));
  }
}
