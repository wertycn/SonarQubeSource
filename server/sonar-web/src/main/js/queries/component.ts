/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { UseQueryResult, useQuery } from '@tanstack/react-query';
import { getTasksForComponent } from '../api/ce';
import { getMeasuresWithMetrics } from '../api/measures';
import { BranchParameters } from '../types/branch-like';
import { MeasuresAndMetaWithMetrics } from '../types/measures';
import { Component } from '../types/types';

const TASK_RETRY = 10_000;

type QueryKeyData = {
  metricKeys: string[];
  branchParameters: BranchParameters;
};

function getComponentQueryKey(key: string, type: 'tasks'): string[];
function getComponentQueryKey(key: string, type: 'measures', data: QueryKeyData): string[];
function getComponentQueryKey(key: string, type: string, data?: QueryKeyData): string[] {
  return ['component', key, type, JSON.stringify(data)];
}

function extractQueryKeyData(queryKey: string[]): { key: string; data?: QueryKeyData } {
  const [, key, , data] = queryKey;
  return { key, data: JSON.parse(data ?? 'null') };
}

export function useTaskForComponentQuery(component: Component) {
  return useQuery({
    queryKey: getComponentQueryKey(component.key, 'tasks'),
    queryFn: ({ queryKey }) => {
      const { key } = extractQueryKeyData(queryKey);
      return getTasksForComponent(key);
    },
    refetchInterval: TASK_RETRY,
  });
}

export function useComponentMeasuresWithMetricsQuery(
  key: string,
  metricKeys: string[],
  branchParameters: BranchParameters,
  enabled = true,
): UseQueryResult<MeasuresAndMetaWithMetrics> {
  return useQuery({
    enabled,
    queryKey: getComponentQueryKey(key, 'measures', {
      metricKeys,
      branchParameters,
    }),
    queryFn: ({ queryKey }) => {
      const { key, data } = extractQueryKeyData(queryKey);
      return data && getMeasuresWithMetrics(key, data.metricKeys, data.branchParameters);
    },
  });
}
