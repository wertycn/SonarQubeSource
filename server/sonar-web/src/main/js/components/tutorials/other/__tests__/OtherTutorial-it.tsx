/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import userEvent from '@testing-library/user-event';
import React from 'react';
import selectEvent from 'react-select-event';
import UserTokensMock from '../../../../api/mocks/UserTokensMock';
import { mockComponent } from '../../../../helpers/mocks/component';
import { mockLanguage, mockLoggedInUser } from '../../../../helpers/testMocks';
import { RenderContext, renderApp } from '../../../../helpers/testReactTestingUtils';
import { byRole, byText } from '../../../../helpers/testSelector';
import {
  getCopyToClipboardValue,
  getTutorialActionButtons,
  getTutorialBuildButtons,
} from '../../test-utils';
import OtherTutorial from '../OtherTutorial';

jest.mock('../../../../api/settings', () => ({
  getAllValues: jest.fn().mockResolvedValue([]),
}));

const tokenMock = new UserTokensMock();

afterEach(() => {
  tokenMock.reset();
});

const ui = {
  provideTokenTitle: byRole('heading', { name: 'onboarding.token.header' }),
  runAnalysisTitle: byRole('heading', { name: 'onboarding.analysis.header' }),
  generateTokenRadio: byRole('radio', { name: 'onboarding.token.generate.PROJECT_ANALYSIS_TOKEN' }),
  existingTokenRadio: byRole('radio', { name: 'onboarding.token.use_existing_token' }),
  tokenNameInput: byRole('textbox', { name: /onboarding.token.name.label/ }),
  expiresInSelect: byRole('combobox', { name: '' }),
  tokenValueInput: byRole('textbox', { name: /onboarding.token.use_existing_token.label/ }),
  invalidTokenValueMessage: byText('onboarding.token.invalid_format'),
  ...getTutorialActionButtons(),
  ...getTutorialBuildButtons(),
};

it('should generate/delete a new token or use existing one', async () => {
  const user = userEvent.setup();
  renderOtherTutorial();

  // Verify that pages is rendered and includes 2 steps
  expect(await ui.provideTokenTitle.find()).toBeInTheDocument();
  expect(ui.runAnalysisTitle.get()).toBeInTheDocument();

  // Generating token
  await user.type(ui.tokenNameInput.get(), 'Testing token');
  await selectEvent.select(ui.expiresInSelect.get(), 'users.tokens.expiration.365');
  await user.click(ui.generateTokenButton.get());

  expect(ui.continueButton.get()).toBeEnabled();

  // Deleting generated token & switchning to existing one
  await user.click(ui.deleteTokenButton.get());

  await user.click(ui.existingTokenRadio.get());
  await user.type(ui.tokenValueInput.get(), 'INVALID TOKEN VALUE');
  expect(ui.invalidTokenValueMessage.get()).toBeInTheDocument();

  await user.clear(ui.tokenValueInput.get());
  await user.type(ui.tokenValueInput.get(), 'validtokenvalue');
  expect(ui.continueButton.get()).toBeEnabled();

  // navigate to 'Run analysis' step
  await user.click(ui.continueButton.get());
  expect(ui.describeBuildTitle.get()).toBeInTheDocument();

  // navigate to previous step
  await user.click(ui.provideTokenTitle.get());
  expect(ui.continueButton.get()).toBeEnabled();
});

it('can choose build tools and copy provided settings', async () => {
  const user = userEvent.setup();
  renderOtherTutorial();

  await user.click(ui.generateTokenButton.get());
  await user.click(ui.continueButton.get());

  // Maven
  await user.click(ui.mavenBuildButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('maven: execute scanner');

  // Gradle
  await user.click(ui.gradleBuildButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('gradle: sonarqube plugin');
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('gradle: execute scanner');

  // Dotnet - Core
  await user.click(ui.dotnetBuildButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot(
    'dotnet core: install scanner globally',
  );
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('dotnet core: execute command 1');
  expect(getCopyToClipboardValue(2, 'Copy')).toMatchSnapshot('dotnet core: execute command 2');
  expect(getCopyToClipboardValue(3, 'Copy')).toMatchSnapshot('dotnet core: execute command 3');

  // Dotnet - Framework
  await user.click(ui.dotnetFrameworkButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('dotnet framework: execute command 1');
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('dotnet framework: execute command 2');
  expect(getCopyToClipboardValue(2, 'Copy')).toMatchSnapshot('dotnet framework: execute command 3');

  // C Family - Linux
  await user.click(ui.cFamilyBuildButton.get());
  await user.click(ui.linuxButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot(
    'cfamily linux: execute build wrapper',
  );
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('cfamily linux: execute scanner');

  // C Family - Windows
  await user.click(ui.windowsButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot(
    'cfamily windows: execute build wrapper',
  );
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('cfamily windows: execute scanner');

  // C Family - MacOS
  await user.click(ui.macosButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot(
    'cfamily macos: execute build wrapper',
  );
  expect(getCopyToClipboardValue(1, 'Copy')).toMatchSnapshot('cfamily macos: execute scanner');

  // Other - Linux
  await user.click(ui.otherBuildButton.get());
  await user.click(ui.linuxButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('other linux: execute scanner');

  // Other - Windows
  await user.click(ui.windowsButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('other windows: execute scanner');

  // Other - MacOS
  await user.click(ui.macosButton.get());
  expect(getCopyToClipboardValue(0, 'Copy')).toMatchSnapshot('other macos: execute scanner');
});

function renderOtherTutorial({
  languages = { c: mockLanguage({ key: 'c' }) },
}: RenderContext = {}) {
  return renderApp(
    '/',
    <OtherTutorial
      baseUrl="http://localhost:9000"
      component={mockComponent()}
      currentUser={mockLoggedInUser()}
    />,
    { languages },
  );
}
