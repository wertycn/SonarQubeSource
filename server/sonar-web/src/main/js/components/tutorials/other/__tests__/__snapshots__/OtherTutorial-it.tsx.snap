// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can choose build tools and copy provided settings: cfamily linux: execute build wrapper 1`] = `
"curl --create-dirs -sSLo $HOME/.sonar/build-wrapper-linux-x86.zip http://localhost:9000/static/cpp/build-wrapper-linux-x86.zip
unzip -o $HOME/.sonar/build-wrapper-linux-x86.zip -d $HOME/.sonar/
export PATH=$HOME/.sonar/build-wrapper-linux-x86:$PATH
"
`;

exports[`can choose build tools and copy provided settings: cfamily linux: execute scanner 1`] = `
"export SONAR_SCANNER_VERSION=5.0.1.3006
export SONAR_SCANNER_HOME=$HOME/.sonar/sonar-scanner-$SONAR_SCANNER_VERSION-linux
curl --create-dirs -sSLo $HOME/.sonar/sonar-scanner.zip https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$SONAR_SCANNER_VERSION-linux.zip
unzip -o $HOME/.sonar/sonar-scanner.zip -d $HOME/.sonar/
export PATH=$SONAR_SCANNER_HOME/bin:$PATH
export SONAR_SCANNER_OPTS="-server"
"
`;

exports[`can choose build tools and copy provided settings: cfamily macos: execute build wrapper 1`] = `
"curl --create-dirs -sSLo $HOME/.sonar/build-wrapper-macosx-x86.zip http://localhost:9000/static/cpp/build-wrapper-macosx-x86.zip
unzip -o $HOME/.sonar/build-wrapper-macosx-x86.zip -d $HOME/.sonar/
export PATH=$HOME/.sonar/build-wrapper-macosx-x86:$PATH
"
`;

exports[`can choose build tools and copy provided settings: cfamily macos: execute scanner 1`] = `
"export SONAR_SCANNER_VERSION=5.0.1.3006
export SONAR_SCANNER_HOME=$HOME/.sonar/sonar-scanner-$SONAR_SCANNER_VERSION-macosx
curl --create-dirs -sSLo $HOME/.sonar/sonar-scanner.zip https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$SONAR_SCANNER_VERSION-macosx.zip
unzip -o $HOME/.sonar/sonar-scanner.zip -d $HOME/.sonar/
export PATH=$SONAR_SCANNER_HOME/bin:$PATH
export SONAR_SCANNER_OPTS="-server"
"
`;

exports[`can choose build tools and copy provided settings: cfamily windows: execute build wrapper 1`] = `
"$env:SONAR_DIRECTORY = [System.IO.Path]::Combine($(get-location).Path,".sonar")
rm "$env:SONAR_DIRECTORY/build-wrapper-win-x86" -Force -Recurse -ErrorAction SilentlyContinue
New-Item -path $env:SONAR_DIRECTORY/build-wrapper-win-x86 -type directory
(New-Object System.Net.WebClient).DownloadFile("http://localhost:9000/static/cpp/build-wrapper-win-x86.zip", "$env:SONAR_DIRECTORY/build-wrapper-win-x86.zip")
Add-Type -AssemblyName System.IO.Compression.FileSystem
[System.IO.Compression.ZipFile]::ExtractToDirectory("$env:SONAR_DIRECTORY/build-wrapper-win-x86.zip", "$env:SONAR_DIRECTORY")
$env:Path += ";$env:SONAR_DIRECTORY/build-wrapper-win-x86"
"
`;

exports[`can choose build tools and copy provided settings: cfamily windows: execute scanner 1`] = `
"$env:SONAR_SCANNER_VERSION = "5.0.1.3006"
$env:SONAR_DIRECTORY = [System.IO.Path]::Combine($(get-location).Path,".sonar")
$env:SONAR_SCANNER_HOME = "$env:SONAR_DIRECTORY/sonar-scanner-$env:SONAR_SCANNER_VERSION-windows"
rm $env:SONAR_SCANNER_HOME -Force -Recurse -ErrorAction SilentlyContinue
New-Item -path $env:SONAR_SCANNER_HOME -type directory
(New-Object System.Net.WebClient).DownloadFile("https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$env:SONAR_SCANNER_VERSION-windows.zip", "$env:SONAR_DIRECTORY/sonar-scanner.zip")
Add-Type -AssemblyName System.IO.Compression.FileSystem
[System.IO.Compression.ZipFile]::ExtractToDirectory("$env:SONAR_DIRECTORY/sonar-scanner.zip", "$env:SONAR_DIRECTORY")
rm ./.sonar/sonar-scanner.zip -Force -ErrorAction SilentlyContinue
$env:Path += ";$env:SONAR_SCANNER_HOME/bin"
$env:SONAR_SCANNER_OPTS="-server"
"
`;

exports[`can choose build tools and copy provided settings: dotnet core: execute command 1 1`] = `"dotnet sonarscanner begin /k:"my-project" /d:sonar.host.url="http://localhost:9000"  /d:sonar.token="generatedtoken2""`;

exports[`can choose build tools and copy provided settings: dotnet core: execute command 2 1`] = `"dotnet build"`;

exports[`can choose build tools and copy provided settings: dotnet core: execute command 3 1`] = `"dotnet sonarscanner end /d:sonar.token="generatedtoken2""`;

exports[`can choose build tools and copy provided settings: dotnet core: install scanner globally 1`] = `"dotnet tool install --global dotnet-sonarscanner"`;

exports[`can choose build tools and copy provided settings: dotnet framework: execute command 1 1`] = `"SonarScanner.MSBuild.exe begin /k:"my-project" /d:sonar.host.url="http://localhost:9000" /d:sonar.token="generatedtoken2""`;

exports[`can choose build tools and copy provided settings: dotnet framework: execute command 2 1`] = `"MsBuild.exe /t:Rebuild"`;

exports[`can choose build tools and copy provided settings: dotnet framework: execute command 3 1`] = `"SonarScanner.MSBuild.exe end /d:sonar.token="generatedtoken2""`;

exports[`can choose build tools and copy provided settings: gradle: execute scanner 1`] = `
"./gradlew sonar \\
  -Dsonar.projectKey=my-project \\
  -Dsonar.projectName='MyProject' \\
  -Dsonar.host.url=http://localhost:9000 \\
  -Dsonar.token=generatedtoken2"
`;

exports[`can choose build tools and copy provided settings: gradle: sonarqube plugin 1`] = `
"plugins {
  id "org.sonarqube" version "4.4.1.3373"
}"
`;

exports[`can choose build tools and copy provided settings: maven: execute scanner 1`] = `
"mvn clean verify sonar:sonar \\
  -Dsonar.projectKey=my-project \\
  -Dsonar.projectName='MyProject' \\
  -Dsonar.host.url=http://localhost:9000 \\
  -Dsonar.token=generatedtoken2"
`;

exports[`can choose build tools and copy provided settings: other linux: execute scanner 1`] = `
"export SONAR_SCANNER_VERSION=5.0.1.3006
export SONAR_SCANNER_HOME=$HOME/.sonar/sonar-scanner-$SONAR_SCANNER_VERSION-linux
curl --create-dirs -sSLo $HOME/.sonar/sonar-scanner.zip https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$SONAR_SCANNER_VERSION-linux.zip
unzip -o $HOME/.sonar/sonar-scanner.zip -d $HOME/.sonar/
export PATH=$SONAR_SCANNER_HOME/bin:$PATH
export SONAR_SCANNER_OPTS="-server"
"
`;

exports[`can choose build tools and copy provided settings: other macos: execute scanner 1`] = `
"export SONAR_SCANNER_VERSION=5.0.1.3006
export SONAR_SCANNER_HOME=$HOME/.sonar/sonar-scanner-$SONAR_SCANNER_VERSION-macosx
curl --create-dirs -sSLo $HOME/.sonar/sonar-scanner.zip https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$SONAR_SCANNER_VERSION-macosx.zip
unzip -o $HOME/.sonar/sonar-scanner.zip -d $HOME/.sonar/
export PATH=$SONAR_SCANNER_HOME/bin:$PATH
export SONAR_SCANNER_OPTS="-server"
"
`;

exports[`can choose build tools and copy provided settings: other windows: execute scanner 1`] = `
"$env:SONAR_SCANNER_VERSION = "5.0.1.3006"
$env:SONAR_DIRECTORY = [System.IO.Path]::Combine($(get-location).Path,".sonar")
$env:SONAR_SCANNER_HOME = "$env:SONAR_DIRECTORY/sonar-scanner-$env:SONAR_SCANNER_VERSION-windows"
rm $env:SONAR_SCANNER_HOME -Force -Recurse -ErrorAction SilentlyContinue
New-Item -path $env:SONAR_SCANNER_HOME -type directory
(New-Object System.Net.WebClient).DownloadFile("https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-$env:SONAR_SCANNER_VERSION-windows.zip", "$env:SONAR_DIRECTORY/sonar-scanner.zip")
Add-Type -AssemblyName System.IO.Compression.FileSystem
[System.IO.Compression.ZipFile]::ExtractToDirectory("$env:SONAR_DIRECTORY/sonar-scanner.zip", "$env:SONAR_DIRECTORY")
rm ./.sonar/sonar-scanner.zip -Force -ErrorAction SilentlyContinue
$env:Path += ";$env:SONAR_SCANNER_HOME/bin"
$env:SONAR_SCANNER_OPTS="-server"
"
`;
